File: python/src/dsa/data_structures/queue.py
"""Queue implementation using collections.deque.

Time: O(1) for enqueue, dequeue, peek
Space: O(n)
"""

from collections import deque
from typing import Generic, TypeVar, Optional

T = TypeVar("T")


class Queue(Generic[T]):
    """A generic queue implementation."""

    def __init__(self) -> None:
        self._data: deque[T] = deque()

    def enqueue(self, value: T) -> None:
        """Add a value to the back of the queue."""
        self._data.append(value)

    def dequeue(self) -> Optional[T]:
        """Remove and return the front value, or None if empty."""
        if self._data:
            return self._data.popleft()
        return None

    def peek(self) -> Optional[T]:
        """Return the front value without removing it, or None if empty."""
        if self._data:
            return self._data[0]
        return None

    def is_empty(self) -> bool:
        """Return True if the queue is empty."""
        return len(self._data) == 0

    def __len__(self) -> int:
        """Return the number of elements in the queue."""
        return len(self._data)

    def clear(self) -> None:
        """Remove all elements from the queue."""
        self._data.clear()

File: nim/src/dsa/data_structures/queue.nim
## Queue - FIFO data structure
##
## Time Complexity:
## - enqueue: O(1) amortized
## - dequeue: O(1) amortized
## - peek: O(1)
## - isEmpty: O(1)
## - size: O(1)

type
  Queue*[T] = object
    items: seq[T]
    front: int

proc newQueue*[T](): Queue[T] =
  ## Create a new empty queue
  Queue[T](items: @[], front: 0)

proc enqueue*[T](q: var Queue[T], item: T) =
  ## Add an item to the back of the queue
  q.items.add(item)

proc dequeue*[T](q: var Queue[T]): T =
  ## Remove and return the front item
  ## Raises IndexDefect if queue is empty
  if q.front >= q.items.len:
    raise newException(IndexDefect, "Queue is empty")
  result = q.items[q.front]
  inc q.front
  # Compact if too much wasted space
  if q.front > q.items.len div 2 and q.front > 16:
    q.items = q.items[q.front .. ^1]
    q.front = 0

proc peek*[T](q: Queue[T]): T =
  ## Return the front item without removing it
  ## Raises IndexDefect if queue is empty
  if q.front >= q.items.len:
    raise newException(IndexDefect, "Queue is empty")
  q.items[q.front]

proc isEmpty*[T](q: Queue[T]): bool =
  ## Check if the queue is empty
  q.front >= q.items.len

proc size*[T](q: Queue[T]): int =
  ## Return the number of items in the queue
  q.items.len - q.front

proc clear*[T](q: var Queue[T]) =
  ## Remove all items from the queue
  q.items.setLen(0)
  q.front = 0

proc toSeq*[T](q: Queue[T]): seq[T] =
  ## Convert queue to sequence (front to back)
  if q.front >= q.items.len:
    @[]
  else:
    q.items[q.front .. ^1]

when isMainModule:
  var q = newQueue[int]()
  q.enqueue(1)
  q.enqueue(2)
  q.enqueue(3)
  assert q.dequeue() == 1
  assert q.peek() == 2
  assert q.size() == 2
  assert not q.isEmpty()
  echo "Queue tests passed!"
